nat/service_ics.go

```diff
@@ -0,0 +1,147 @@
+	if err != nil {
+		return err
+	}
+
+	// Enable internet connection sharing for the public interface.
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.EnableSharing(0)`)
+
+	return err
+}
+
+func (nat *serviceICS) Add(rule RuleForwarding) error {
+	// Enable internet connection sharing for the local interface.
+	// TODO detect interface name from the rule.
+	// TODO firewall rule configuration should be added here for new connections.
+	ifaceName := "myst1"
+
+	_, err := powerShell(`regsvr32 /s hnetcfg.dll;
+	$netShare = New-Object -ComObject HNetCfg.HNetShare;
+	$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+	$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+	$config.EnableSharing(1)`)
+
+	return err
+}
+
+func (nat *serviceICS) Del(rule RuleForwarding) error {
+	// Disable internet connection sharing for the local interface.
+	// TODO detect interace name from the rule.
+	// TODO firewall rule configuration should be added here for cleaning up unused connections.
+	ifaceName := "myst1"
+
+	_, err := powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+
+	return err
+}
+
+func (nat *serviceICS) Disable() error {
+	// TODO stop internet connection sharing on all enabled interfaces (nat.ifaces).
+	// TODO we should cleanup as much as possible, not failing after errors.
+	_, err := powerShell("Set-Service -Name RemoteAccess -StartupType " + nat.remoteAccessStatus)
+	if err != nil {
+		return err
+	}
+
+	ifaceName, err := getPublicInterfaceName()
+	if err != nil {
+		return err
+	}
+
+	// Disable internet connection sharing for the public interface.
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+
+	return err
+}
+
+func getPublicInterfaceName() (string, error) {
+	out, err := powerShell(`Get-WmiObject -Class Win32_IP4RouteTable | where { $_.destination -eq '0.0.0.0' -and $_.mask -eq '0.0.0.0'} | foreach { $_.InterfaceIndex }`)
+	if err != nil {
+		return "", err
+	}
+
+	ifaceID, err := strconv.Atoi(strings.TrimSpace(string(out)))
+	if err != nil {
+		return "", err
+
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return "", err
```

Comment:
We could use errors.warp here pinpoint more specific place with "failed to fetch system interfaces" or so. Same above.